File: bot-java/yfinance-rest/src/main/java/com/yahoo/finanance/query1/Quote.java
package com.yahoo.finanance.query1;

import com.dslplatform.json.CompiledJson;
import com.dslplatform.json.JsonAttribute;
import com.dslplatform.json.TimestampConverter;
import java.text.DecimalFormat;
import java.util.Calendar;
import java.util.Optional;
import java.util.function.Function;
import lombok.Data;
import lombok.ToString;

@Data
@ToString
@CompiledJson
public class Quote {
  public enum QuoteType {
    ETF,
    EQUITY
  }

  private String language;
  private String region;
  private String quoteType;
  private String quoteSourceName;
  private boolean triggerable;
  private String currency;
  private String marketState;
  private String exchange;
  private String longName;
  private String messageBoardId;
  private String exchangeTimezoneName;
  private String market;
  private long firstTradeDateMilliseconds;
  private int priceHint;
  private double postMarketChangePercent;
  private int postMarketTime;
  private double postMarketPrice;
  private double postMarketChange;
  private double regularMarketChange;
  private double regularMarketChangePercent;
  private int regularMarketTime;
  private double regularMarketPrice;
  private double regularMarketDayHigh;
  private String regularMarketDayRange;
  private double regularMarketDayLow;
  private int regularMarketVolume;
  private double regularMarketPreviousClose;
  private double bid;
  private double ask;
  private int bidSize;
  private int askSize;
  private String fullExchangeName;
  private String financialCurrency;
  private double regularMarketOpen;
  private int averageDailyVolume3Month;
  private int averageDailyVolume10Day;
  private double fiftyTwoWeekLowChange;
  private double fiftyTwoWeekLowChangePercent;
  private String fiftyTwoWeekRange;
  private double fiftyTwoWeekHighChange;
  private double fiftyTwoWeekHighChangePercent;
  private double fiftyTwoWeekLow;
  private double fiftyTwoWeekHigh;

  @JsonAttribute(converter = TimestampConverter.class)
  private Calendar dividendDate;

  @JsonAttribute(converter = TimestampConverter.class)
  private Calendar earningsTimestamp;

  @JsonAttribute(converter = TimestampConverter.class)
  private Calendar earningsTimestampStart;

  @JsonAttribute(converter = TimestampConverter.class)
  private Calendar earningsTimestampEnd;

  private double trailingAnnualDividendRate;
  private double trailingPE;
  private double trailingAnnualDividendYield;
  private double epsTrailingTwelveMonths;
  private double epsForward;
  private double epsCurrentYear;
  private double priceEpsCurrentYear;
  private long sharesOutstanding;
  private double bookValue;
  private double fiftyDayAverage;
  private double fiftyDayAverageChange;
  private double fiftyDayAverageChangePercent;
  private double twoHundredDayAverage;
  private double twoHundredDayAverageChange;
  private double twoHundredDayAverageChangePercent;
  private long marketCap;
  private double forwardPE;
  private double priceToBook;
  private int sourceInterval;
  private int exchangeDataDelayedBy;
  private String averageAnalystRating;
  private boolean tradeable;
  private String shortName;
  private String exchangeTimezoneShortName;
  private int gmtOffSetMilliseconds;
  private boolean esgPopulated;
  private String displayName;
  private String symbol;

  private String expenseRatio;
  private String aum;

  public static Function<Calendar, String> calendar2str() {
    return time -> Optional.ofNullable(time)
        .map(x -> x.getTime().toString())
        .orElse("");
  }

  public Function<Object, Object> dividendDateStr() {
    return x -> calendar2str().apply(dividendDate);
  }

  public Function<Object, Object> earningsTimestampStr() {
    return x -> calendar2str().apply(earningsTimestamp);
  }

  public Function<Object, Object> trailingAnnualDividendYieldStr() {
    return x -> new DecimalFormat("#0.00%").format(trailingAnnualDividendYield);
  }
}

File: bot-scala/src/main/scala/com/yahoo/finanance/query1/Quote.scala
package com.yahoo.finanance.query1

import com.github.mustachejava.{DefaultMustacheFactory, Mustache}
import io.circe._
import io.circe.generic.semiauto._

import java.io.StringWriter
import java.time.{LocalDateTime, ZoneId}
import java.util.{Calendar, Date}
import scala.util.{Try, Using}

case class Quote(
                  language: String,
                  region: String,
                  quoteType: String,
                  quoteSourceName: String,
                  triggerable: Boolean,
                  currency: String,
                  marketState: String,
                  exchange: String,
                  longName: String,
                  messageBoardId: String,
                  exchangeTimezoneName: String,
                  market: String,
                  firstTradeDateMilliseconds: Long,
                  priceHint: Int,
                  postMarketChangePercent: Option[Double],
                  postMarketTime: Option[Int],
                  postMarketPrice: Option[Double],
                  postMarketChange: Option[Double],
                  regularMarketChange: Double,
                  regularMarketChangePercent: Double,
                  regularMarketTime: Int,
                  regularMarketPrice: Double,
                  regularMarketDayHigh: Double,
                  regularMarketDayRange: String,
                  regularMarketDayLow: Double,
                  regularMarketVolume: Int,
                  regularMarketPreviousClose: Double,
                  bid: Double,
                  ask: Double,
                  bidSize: Int,
                  askSize: Int,
                  fullExchangeName: String,
                  financialCurrency: Option[String],
                  regularMarketOpen: Double,
                  averageDailyVolume3Month: Int,
                  averageDailyVolume10Day: Int,
                  fiftyTwoWeekLowChange: Double,
                  fiftyTwoWeekLowChangePercent: Double,
                  fiftyTwoWeekRange: String,
                  fiftyTwoWeekHighChange: Double,
                  fiftyTwoWeekHighChangePercent: Double,
                  fiftyTwoWeekLow: Double,
                  fiftyTwoWeekHigh: Double,
                  dividendDate: Option[Calendar],
                  earningsTimestamp: Option[Calendar],
                  earningsTimestampStart: Option[Calendar],
                  earningsTimestampEnd: Option[Calendar],
                  trailingAnnualDividendRate: Option[Double],
                  trailingPE: Option[Double],
                  trailingAnnualDividendYield: Option[Double],
                  epsTrailingTwelveMonths: Option[Double],
                  epsForward: Option[Double],
                  epsCurrentYear: Option[Double],
                  priceEpsCurrentYear: Option[Double],
                  sharesOutstanding: Option[Long],
                  bookValue: Option[Double],
                  fiftyDayAverage: Double,
                  fiftyDayAverageChange: Double,
                  fiftyDayAverageChangePercent: Double,
                  twoHundredDayAverage: Double,
                  twoHundredDayAverageChange: Double,
                  twoHundredDayAverageChangePercent: Double,
                  marketCap: Option[Long],
                  forwardPE: Option[Double],
                  priceToBook: Option[Double],
                  sourceInterval: Int,
                  exchangeDataDelayedBy: Int,
                  averageAnalystRating: Option[String],
                  tradeable: Boolean,
                  shortName: String,
                  exchangeTimezoneShortName: String,
                  gmtOffSetMilliseconds: Int,
                  esgPopulated: Boolean,
                  displayName: Option[String],
                  symbol: String,
                  expenseRatio: Option[String],
                  aum: Option[String]
                )

object Quote {
  implicit val encoder: Encoder[Quote] = deriveEncoder[Quote]
  implicit val decoder: Decoder[Quote] = deriveDecoder[Quote]

  implicit val calendarEncoder: Encoder[Calendar] =
    (date: Calendar) => Json.fromLong(timestamp(date))

  implicit val calendarDecoder: Decoder[Calendar] =
    (c: HCursor) => c.as[Long].map(x => calendar(x))

  def calendar(timestamp: Long): Calendar = {
    val calendar = Calendar.getInstance
    calendar.setTimeInMillis(timestamp * 1000)
    calendar
  }

  def timestamp(calendar: Calendar): Long = calendar.getTimeInMillis / 1000

  def timestamp(time: LocalDateTime): Long = Date.from(time.atZone(ZoneId.systemDefault).toInstant).getTime / 1000

  lazy val mf = new DefaultMustacheFactory
  lazy val quoteMustache: Mustache = mf.compile("quote.mustache")
  lazy val errorMustache: Mustache = mf.compile("error.mustache")

  implicit class QuoteHelper(q: Quote) {
    def copyWith(expenseRatio: String, aum: String): Quote = q.copy(
      expenseRatio = Option(expenseRatio),
      aum = Option(aum))

    def html: Try[String] = Using(new StringWriter())(w => {
      quoteMustache.execute(w, q).flush()
      w.toString
    })
  }

  implicit class ThrowableHelper(err: Throwable) {
    def html: Try[String] = Using(new StringWriter())(w => {
      errorMustache.execute(w, err).flush()
      w.toString
    })
  }

}